libs/pilight/protocols/433.92/selectplus_doorbell.c

```diff
@@ -0,0 +1,183 @@
+/*
+	Copyright (C) 2020 woutput
+
+	This file is part of pilight.
+	This Source Code Form is subject to the terms of the Mozilla Public
+	License, v. 2.0. If a copy of the MPL was not distributed with this
+	file, You can obtain one at http://mozilla.org/MPL/2.0/.
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+
+#include "../../core/pilight.h"
+#include "../../core/common.h"
+#include "../../core/dso.h"
+#include "../../core/log.h"
+#include "../protocol.h"
+#include "../../core/binary.h"
+#include "../../core/gc.h"
+#include "selectplus_doorbell.h"
+
+// Timing from TODO
+
+// Increase this value to be more robust, but also create more false positives
+#define PEAK_TO_PEAK_JITTER	80
+
+// The same jitter is assumed on every pulse type (as may be expected)
+
+// Short pulse timing
+#define MIN_SHORT_PULSE_LENGTH	(AVG_SHORT_PULSE_LENGTH - 0.5 * PEAK_TO_PEAK_JITTER)
+#define AVG_SHORT_PULSE_LENGTH	372
+#define MAX_SHORT_PULSE_LENGTH	(AVG_SHORT_PULSE_LENGTH + 0.5 * PEAK_TO_PEAK_JITTER)
+
+// Medium pulse timing
+#define MIN_MEDIUM_PULSE_LENGTH	(AVG_MEDIUM_PULSE_LENGTH - 0.5 * PEAK_TO_PEAK_JITTER)
+#define AVG_MEDIUM_PULSE_LENGTH	1094
+#define MAX_MEDIUM_PULSE_LENGTH	(AVG_MEDIUM_PULSE_LENGTH + 0.5 * PEAK_TO_PEAK_JITTER)
+
+// Long pulse timing
+#define MIN_LONG_PULSE_LENGTH	(AVG_LONG_PULSE_LENGTH - 0.5 * PEAK_TO_PEAK_JITTER)
+#define AVG_LONG_PULSE_LENGTH	6536
+#define MAX_LONG_PULSE_LENGTH	(AVG_LONG_PULSE_LENGTH + 0.5 * PEAK_TO_PEAK_JITTER)
+
+#define RAW_LENGTH		36
+// One pulse per bit, first 27 pulses are ID, last 9 pulses are footer
+#define BINARY_LENGTH		27
+
+#define NORMAL_REPEATS		68
+
+static int validate(void) {
+	// Check for match in raw length
+	if (selectplus_doorbell->rawlen == RAW_LENGTH) {
+		// Check for match in only part of footer
+		if ((selectplus_doorbell->raw[34] >= MIN_MEDIUM_PULSE_LENGTH) &&
+		    (selectplus_doorbell->raw[34] <= MAX_MEDIUM_PULSE_LENGTH) &&
+		    (selectplus_doorbell->raw[35] >= MIN_LONG_PULSE_LENGTH) &&
+		    (selectplus_doorbell->raw[35] <= MAX_LONG_PULSE_LENGTH)) {
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+static void createMessage(int id) {
+	selectplus_doorbell->message = json_mkobject();
+	json_append_member(selectplus_doorbell->message, "id", json_mknumber(id, 0));
+	selectplus_doorbell->txrpt = NORMAL_REPEATS;
+}
+
+static void parseCode(void) {
+	int binary[BINARY_LENGTH], x;
+
+	for (x = 0; x < BINARY_LENGTH; x += 1) {
+		if ((selectplus_doorbell->raw[x] >= MIN_SHORT_PULSE_LENGTH) &&
```

Comment:
Can you follow pilight styleguide as much as possible?

Reply:
Fixed (?)